### ISSUE TITLE
[Extension Proposal] Docker Java Client

### ISSUE THREAD

**Issue by zZHorizonZz** [quarkusio/quarkus#44255]
### Description

I have been working on a Quarkus Docker Java client for one of my projects. Since it's separate, I have decided to split it and make it into its own extension.

This extension adds the ability to configure and inject DockerClient from [Docker Java Client](https://github.com/docker-java/docker-java). Also, if no host is configured, the host is configured automatically depending on the user's OS. If the user is using Unix, the `unix:///var/run/docker.sock` will be used if the user is on Windows, then `npipe:////./pipe/docker_engine` is used.

You can use the extension like this:
```java
@Inject
DockerClient dockerClient;

// Use the client
dockerClient.pingCmd().exec();
```

Or you can create a named client like this:

```java
@Inject
@NamedDockerClient("production")
DockerClient productionClient;
```

With configuration being like this:

```
# Production client configuration
quarkus.docker."production".docker-host=tcp://prod-host:2375
quarkus.docker."production".enabled=true
```

Currently, the extension doesn't support native mode as there are problems with using sockets in the library default http client. In the future, I think there could be a custom VertX implementation of this client, similar to how Fabric8 does it. The extension also provides a readiness health check to see if all clients started properly.

### Repository name

quarkus-docker-client

### Short description

Docker Java client integration for Quarkus

### Repository Homepage URL

https://github.com/zZHorizonZz/quarkus-docker-client

### Repository Topics

- quarkus-extension
- docker
- container
- docker-api
- containerization

### Team Members

- @zZHorizonZz (Daniel Fiala) me

**Comment by maxandersen** [comment]
I needed this today :) seems cool to have imo.

**Comment by zZHorizonZz** [comment]
I would like to contribute it as a Quarkiverse extension, although obviously there is the issue with it not really working on native (for now). Or I could try to publish it myself, but I didn't really find any guide on how to self publish extensions.

**Comment by maxandersen** [comment]
yeah it would be great to have - i started a conversation on how where it would make most sense [#dev > quarkus docker client](https://quarkusio.zulipchat.com/#narrow/channel/187038-dev/topic/quarkus.20docker.20client)

but for sure lets find a way.

about self publishing it is just a matter of publishing the maven artifacts and hook into the registry - https://quarkus.io/guides/writing-extensions#ecosystem

with quarkiverse 90% of that is done for you automatically - but all doable manually too. if unclear we definitely welcome suggestions on how to make it clearer.

**Comment by maxandersen** [comment]
ok so given the limitations on docker-java's streaming and native integration issues lets put it in quarkiverse if you are up for it.

@gastaldi can you kick it off ?

**Comment by gastaldi** [comment]
Sure, on it

**Comment by gastaldi** [comment]
https://github.com/quarkiverse/quarkus-docker-client is now available, enjoy!

**Comment by zZHorizonZz** [comment]
@maxandersen I have updated some dependencies and made a couple of changes, but since this is my first extension, I want to be sure that everything is as it should be. I have gone through the onboarding and releasing wiki, but could you check if it has everything necessary for the release?

**Comment by maxandersen** [comment]
damn  - i missed the name was docker-client; should we call this docker-java as its for docker-java library?

**Comment by gastaldi** [comment]
I think `quarkus-docker-client` is fine. I'd use `io.quarkiverse.docker` as the groupId though, but that's my personal preference
